### PR TITLE
Add internal user listing

### PR DESCRIPTION
**Pull request by yuxitang-amzn**
*Issue #, if available:*

*Description of changes:*
Add internal user listing.

TODO: 'Current' user badge.

Screenshots:
Mockup:
image

Implementation:
image

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: public/apps/configuration/app-router.tsx [hunk 1]
```diff
@@ -64,7 +65,7 @@ export function AppRouter(props: AppDependencies) {
   return (
     <Router basename={props.params.appBasePath}>
       <EuiPage>
-        <Route path={`(${RoutesList.map((r) => r.href).join('|')})`} exact>
+        <Route path={`(${RoutesList.map((r) => r.href).join('|')}|/)`} exact>
```
Comment (seraphjiang) [inline public/apps/configuration/app-router.tsx:68]: not easy to understand the arrow function, would you create function for readability

File: public/apps/configuration/utils/internal-user-list-utils.tsx [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        username: key,
+        attributes: value.attributes,
+      } as InternalUsersListing)
+  );
+}
+
+export async function requestDeleteUsers(http: HttpStart, users: string[]) {
+  users.forEach(async (r) => {
+    await http.delete(`${API_ENDPOINT_INTERNALUSERS}/${r}`);
+  });
```
Comment (seraphjiang) [inline public/apps/configuration/utils/internal-user-list-utils.tsx:39]: could we check and ensure this is expected way we want here?

https://codeburst.io/javascript-async-await-with-foreach-b6ba62bbf404

Comment (yuxitang-amzn) [reply public/apps/configuration/utils/internal-user-list-utils.tsx:39]: Thanks for pointing out. Tried `for...of` and seems works:
~~~javascript
const delay = ms => new Promise(res => setTimeout(res, ms));
for (let num of [1, 2, 3]) {
  delay(1000);
  console.log(num);
}
console.log("Done");

VM499:4 1
VM499:4 2
VM499:4 3
VM499:6 Done
~~~

Will switch to `for...of`

File: public/apps/configuration/app-router.tsx (hunk 1 above)
Comment (yuxitang-amzn) [reply public/apps/configuration/app-router.tsx]: Created a constant for it.

File: public/apps/configuration/utils/internal-user-list-utils.tsx (hunk 2 above)
Comment (yuxitang-amzn) [reply public/apps/configuration/utils/internal-user-list-utils.tsx:39]: Updated.

File: public/apps/configuration/panels/user-list.tsx [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+                  closePopover={() => {
+                    setActionsPopoverOpen(false);
+                  }}
+                  panelPaddingSize="s"
+                >
+                  <EuiContextMenuPanel items={actionsMenuItems} />
+                </EuiPopover>
+              </EuiFlexItem>
+              <EuiFlexItem>
+                <EuiButton fill>Create user</EuiButton>
```
Comment (tianleh) [inline public/apps/configuration/panels/user-list.tsx:178]: missing "internal"

Comment (yuxitang-amzn) [reply public/apps/configuration/panels/user-list.tsx]: Good catch, fixed.

File: public/apps/configuration/app-router.tsx [hunk 4]
```diff
@@ -13,21 +13,22 @@
...
 import { RoleView } from './panels/role-view/role-view';
+import { UserList } from './panels/user-list';
+import { RouteItem } from './types';
 
-const RoutesMap: { [key: string]: RouteItem } = {
+const ROUTE_MAP: { [key: string]: RouteItem } = {
   getStarted: {
     name: 'Get Started',
-    href: '/getstarted',
+    href: '/',
```
Comment (tianleh) [inline public/apps/configuration/app-router.tsx:31]: looks unrelated to the description of this PR

File: public/apps/configuration/panels/user-list.tsx [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+  }, [props.coreStart.http]);
+
+  const handleDelete = async () => {
+    const usersToDelete: string[] = selection.map((r) => r.username);
+    try {
+      await requestDeleteUsers(props.coreStart.http, usersToDelete);
+      // Refresh from server (calling fetchData) does not work here, the server still return the roles
+      // that had been just deleted, probably because ES takes some time to sync to all nodes.
+      // So here remove the selected roles from local memory directly.
+      setUserData(difference(userData, selection));
```
Comment (tianleh) [inline public/apps/configuration/panels/user-list.tsx:97]: Just want to confirm that if user quickly refreshes the page (before ES deletion is done), that deleted user will show up again.

File: public/apps/configuration/panels/user-list.tsx [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+    <EuiContextMenuItem
+      key="edit"
+      onClick={() => {}} // TODO: Redirect to edit page
+      disabled={selection.length !== 1}
+    >
+      Edit
+    </EuiContextMenuItem>,
+    // TODO: Redirect to duplicate page
+    <EuiContextMenuItem key="duplicate" onClick={() => {}} disabled={selection.length !== 1}>
+      Duplicate
```
Comment (tianleh) [inline public/apps/configuration/panels/user-list.tsx:116]: Looks like the mockup says "Export JSON" instead of "Duplicate"

File: public/apps/configuration/panels/user-list.tsx [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+      <EuiPageContent>
+        <EuiPageContentHeader>
+          <EuiPageContentHeaderSection>
+            <EuiTitle size="s">
+              <h3>Internal users ({userData.length})</h3>
+            </EuiTitle>
+            <EuiText size="xs" color="subdued">
+              The Security plugin includes an internal user database. Use this database in place of
+              or in addition to an external authentication system such as LDAP or Active Directory.
+              You can map an internal user to a role from Roles. First, click into the detail page
```
Comment (tianleh) [inline public/apps/configuration/panels/user-list.tsx:155]: "Roles" looks like a link from the mockup

File: public/apps/configuration/panels/user-list.tsx [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+          </EuiPageContentHeaderSection>
+        </EuiPageContentHeader>
+        <EuiPageBody>
+          <EuiInMemoryTable
+            loading={userData === [] && !errorFlag}
+            columns={columns}
+            items={userData}
+            itemId={'username'}
+            pagination
+            search
```
Comment (tianleh) [inline public/apps/configuration/panels/user-list.tsx:190]: There is a placeholder in the search bar which you can add by the following code

`
  const search = {
    box: {
      placeholder: 'Search internal users',
    },
  };
`

File: public/apps/configuration/panels/user-list.tsx (hunk 6 above)
Comment (yuxitang-amzn) [reply public/apps/configuration/panels/user-list.tsx:116]: Huh, weird, there is a duplicate role page in the mockups:
image

File: public/apps/configuration/panels/user-list.tsx (hunk 5 above)
Comment (yuxitang-amzn) [reply public/apps/configuration/panels/user-list.tsx:97]: In the edge case, yes.

File: public/apps/configuration/utils/internal-user-list-utils.tsx [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+export function transformUserData(rawData: DataObject<InternalUser>): InternalUsersListing[] {
+  return map(rawData, (value: InternalUser, key?: string) => ({
+    username: key || '',
+    attributes: value.attributes,
+  }));
+}
+
+export async function requestDeleteUsers(http: HttpStart, users: string[]) {
+  for (const user of users) {
+    await http.delete(`${API_ENDPOINT_INTERNALUSERS}/${user}`);
```
Comment (tianleh) [inline public/apps/configuration/utils/internal-user-list-utils.tsx:34]: Just curious if one user fails to delete, will it still try to run deletion for the remaining users or stop wherever the first failure happens.

File: public/apps/configuration/panels/user-list.tsx (hunk 7 above)
Comment (yuxitang-amzn) [reply public/apps/configuration/panels/user-list.tsx]: Good catch. Added an empty link.

File: public/apps/configuration/panels/user-list.tsx (hunk 8 above)
Comment (yuxitang-amzn) [reply public/apps/configuration/panels/user-list.tsx]: Didn't notice, updated.

File: public/apps/configuration/utils/internal-user-list-utils.tsx (hunk 9 above)
Comment (yuxitang-amzn) [reply public/apps/configuration/utils/internal-user-list-utils.tsx:34]: It stops upon first failure.

File: public/apps/configuration/panels/user-list.tsx (hunk 6 above)
Comment (tianleh) [reply public/apps/configuration/panels/user-list.tsx:116]: Not sure about the access point for this page. But from the dropdown mockup, this is how it looks like. Probably a follow up to the UX team.

Screen Shot 2020-06-05 at 4 43 33 PM

File: public/apps/configuration/utils/internal-user-list-utils.tsx (hunk 9 above)
Comment (tianleh) [reply public/apps/configuration/utils/internal-user-list-utils.tsx:34]: Seems a better experience to do best effort to delete the remaining. I will leave it to you to decide.

File: public/apps/configuration/panels/user-list.tsx (hunk 6 above)
Comment (yuxitang-amzn) [reply public/apps/configuration/panels/user-list.tsx]: Sure, updated to Export JSON for current stage.

File: public/apps/configuration/utils/internal-user-list-utils.tsx (hunk 9 above)
Comment (yuxitang-amzn) [reply public/apps/configuration/utils/internal-user-list-utils.tsx:34]: In my opinion on the frontend should we keep it simple. It would be better to have a batch deletion API in the future. So I didn't do any extra handling here.
